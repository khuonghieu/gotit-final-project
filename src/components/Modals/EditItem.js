import React, { useState } from 'react';
import { connect } from 'react-redux';
import { Button, Form, Modal } from '@gotitinc/design-system';
import { editItem } from '../../actions/items';

function mapStateToProps(state) {
  return {
    item: state.item,
    categoryId: state.categories.currentCategory,
  };
}

const mapDispatchToProps = (dispatch) => ({
  editItem: (categoryId, itemId, name, description, price) => dispatch(editItem(categoryId, itemId, name, description, price)),
});

export function EditItem({
  onClose, item, editItem, categoryId,
}) {
  const [name, setName] = useState(item.name);
  const [description, setDescription] = useState(item.description);
  const [price, setPrice] = useState(item.price);

  async function handleEditItem(e) {
    e.preventDefault();
    if (name && description && price) {
      const { success, payload } = await editItem(categoryId, item.id, name, description, price);
      if (!success) {
        console.log(payload);
      }
    } else {
      alert('fill all the blanks');
    }
  }

  return (
    <Modal size="small" relative centered onHide={onClose}>
      <Modal.Header closeButton onClick={onClose}>
        <Modal.Title>Edit Item</Modal.Title>
      </Modal.Header>
      <Modal.Body>
        <div className="u-textCenter">
          <img src="holder.js/100px90?text=Image" className="u-maxWidthFull u-marginBottomExtraSmall" alt="" />
        </div>
        <Form.Label>Item name</Form.Label>
        <Form.Input value={name} onChange={(e) => setName(e.target.value)} type="text" placeholder="Enter item name" name="name" />
        <Form.Label>Item description</Form.Label>
        <Form.Input value={description} onChange={(e) => setDescription(e.target.value)} type="text" placeholder="Enter description" name="description" />
        <Form.Label>Item price</Form.Label>
        <Form.Input value={price} onChange={(e) => setPrice(e.target.value)} type="number" placeholder="Enter item price" name="price" />
        <Button onClick={handleEditItem} id="edit-item-form-button">Confirm</Button>
      </Modal.Body>
    </Modal>
  );
}

export default connect(mapStateToProps, mapDispatchToProps)(EditItem);
