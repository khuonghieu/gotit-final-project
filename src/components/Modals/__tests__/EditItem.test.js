import React from 'react';
import { shallow, mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Message, Form, Button } from '@gotitinc/design-system';
import { EditItem } from '../EditItem';

configure({ adapter: new Adapter() });

describe('components/Modals/EditItem.js', () => {
  let wrapper;
  let props;
  let input;
  let button;

  beforeEach(() => {
    props = {
      onClose: jest.fn(),
      prefill: {
        name: 'testname',
        description: 'testdescription',
        id: 0,
        price: 0,
      },
      editItem: jest.fn(),
      categoryId: '0',
    };
  });

  const update = () => {
    input = wrapper.find(Form.Input);
    button = wrapper.find(Button);
  };

  const setup = () => {
    wrapper = shallow(
      <EditItem {...props} />,
    );
    update();
  };

  it('should render correctly', () => {
    props.editItem.mockReturnValue({ success: true, payload: {} });
    setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should call SignIn API when form is fully filled and clicked submit button', () => {
    props.editItem.mockReturnValue({ success: true, payload: {} });
    setup();
    input.at(0).props().onChange({ target: { value: 'testname1' } });
    input.at(1).props().onChange({ target: { value: 'testdescription1' } });
    input.at(2).props().onChange({ target: { value: 5 } });
    update();
    button.props().onClick({ preventDefault: jest.fn() });
    expect(props.editItem).toHaveBeenCalled();
  });

  it('should display error message when all fields are not fully filled', () => {
    props.editItem.mockReturnValue({ success: true, payload: {} });
    setup();
    button.props().onClick({ preventDefault: jest.fn() });
    expect(wrapper.find(Message).length).toBe(1);
  });
  it('should display error message when edit item failed', async () => {
    props.editItem.mockReturnValue({ success: false, payload: { message: 'error' } });
    setup();
    input.at(0).props().onChange({ target: { value: 'testname1' } });
    input.at(1).props().onChange({ target: { value: 'testdescription1' } });
    input.at(2).props().onChange({ target: { value: 5 } });
    update();
    await button.props().onClick({ preventDefault: jest.fn() });
    expect(wrapper.find(Message).length).toBe(1);
  });

  it('should close when edit item success', async () => {
    props.editItem.mockReturnValue({ success: true, payload: {} });
    setup();
    input.at(0).props().onChange({ target: { value: 'testname1' } });
    input.at(1).props().onChange({ target: { value: 'testdescription1' } });
    input.at(2).props().onChange({ target: { value: 5 } });
    update();
    await button.props().onClick({ preventDefault: jest.fn() });
    expect(props.onClose).toHaveBeenCalled();
  });
});
