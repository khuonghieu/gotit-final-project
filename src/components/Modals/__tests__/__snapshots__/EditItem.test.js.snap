// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Modals/EditItem.js should render correctly 1`] = `ReactWrapper {}`;
